### PR TITLE
NIAD-974: Patient Mapper

### PR DESCRIPTION
**Pull request by n-r-b**
## Description

Created `PatientMapper` class which maps the patient in an edifact message to a Fhir Patient resource.

The mapped `Patient` resource has been added to the `PathologyRecord` via the `PathologyRecordMapper`.
The `Patient` in the `PathologyRecord` has been added to the Fhir Bundle as a new entry via the `BundleMapper`.

Integration test data has been updated. The messages in `multi_pathology.dat` are currently duplicates of the first message temporally (until other mappers are in place).

## Jira Ticket

https://gpitbjss.atlassian.net/browse/NIAD-974

## Checklist

These are items (excluding GitHub Checks) which should be confirmed before a branch is ready to merge.

- [x] Acceptance Criteria met
- [x] Commit messages are meaningful
- [x] Manually tested
- [x] Self-reviewed your code
- [ ] Code reviewed from two other developers
- [x] If your pull request depends on any other, please link them in the description
- [x] main branch is passing/green on CI
- [x] Add/update any relevant documentation

### CODE REVIEW

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java [hunk 1]
```diff
@@ -1,12 +1,71 @@
...
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.segmentgroup.InvestigationSubject;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.segmentgroup.PatientDetails;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.segmentgroup.ServiceReportDetails;
+
+import java.util.Optional;
 
 @Component
 public class PatientMapper {
+
+    protected static final String SYSTEM = "https://fhir.nhs.uk/Id/nhs-number";
```
Comment (n-r-b) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java:19]:
```suggestion
    protected static final String NHS_NUMBER_SYSTEM = "https://fhir.nhs.uk/Id/nhs-number";
```

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java [hunk 2]
```diff
@@ -1,12 +1,71 @@
...
+
+        mapGender(patientDetails, patient);
+        mapDateOfBirth(patientDetails, patient);
+
+        return patient;
+    }
+
+    private void mapIdentifier(final PersonName name, final Patient patient) {
+        patient.addIdentifier()
+            .setValue(name.getNhsNumber())
```
Comment (n-r-b) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java:44]: could add a null check for nhsNumber

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java (hunk 1 above)
Comment (n-r-b) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java]: Done

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java (hunk 2 above)
Comment (n-r-b) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java]: Done

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java [hunk 3]
```diff
@@ -1,12 +1,75 @@
...
+        mapDateOfBirth(patientDetails, patient);
+
+        return patient;
+    }
+
+    private void mapIdentifier(final PersonName name, final Patient patient) {
+        final Identifier identifier = new Identifier();
+
+        identifier.setSystem(NHS_NUMBER_SYSTEM);
+        Optional.ofNullable(name.getNhsNumber()).ifPresent(identifier::setValue);
```
Comment (andrewwa-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java:47]: If `getNhsNumber()` returns `null` is there any harm in calling `identifier.setValue(null)`? Just wondering if there's any need to cycle through an `Optional` here

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java [hunk 4]
```diff
@@ -1,12 +1,75 @@
...
+        details.getDateOfBirth()
+            .map(personDateOfBirth -> new DateType(personDateOfBirth.getDateOfBirth()))
+            .ifPresent(patient::setBirthDateElement);
+    }
+
+    private void mapGender(final PatientDetails details, final Patient patient) {
+        details.getSex().ifPresent(personSex -> {
+            final String gender = personSex.getGender().name().toLowerCase();
+            patient.setGender(Enumerations.AdministrativeGender.fromCode(gender));
+        });
```
Comment (andrewwa-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java:62]: Might read better as a chain
```suggestion
        details.getSex()
            .map(sex -> sex.getGender().name().toLowerCase())
            .map(AdministrativeGender::fromCode)
            .ifPresent(patient::setGender);
        });
```

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java [hunk 5]
```diff
@@ -1,12 +1,75 @@
...
+    }
+
+    private void mapName(final PersonName name, final Patient patient) {
+        final HumanName humanName = new HumanName();
+
+        Optional.ofNullable(name.getTitle()).ifPresent(humanName::addPrefix);
+        Optional.ofNullable(name.getFirstForename()).ifPresent(humanName::addGiven);
+        Optional.ofNullable(name.getSecondForename()).ifPresent(humanName::addGiven);
```
Comment (andrewwa-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java:70]: Any chance/risk of second forename being present without a first forename? Would `addGiven` of only the second be a correct mapping in that case?

File: src/test/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapperTest.java [hunk 6]
```diff
@@ -1,19 +1,207 @@
...
+            () -> assertThat(patient.getName()).hasSize(1).first()
+                .satisfies(name -> assertAll(
+                    () -> assertThat(name.getPrefixAsSingleString()).isEqualTo(TITLE),
+                    () -> assertThat(name.getGivenAsSingleString()).isEqualTo(FIRST_FORENAME + " " + SECOND_FORENAME),
+                    () -> assertThat(name.getFamily()).isEqualTo(SURNAME))),
+
+            () -> assertThat(patient.getGender()).isEqualTo(Enumerations.AdministrativeGender.FEMALE),
+
+            () -> {
+                final Date birthDate = Date.from(LocalDate.of(1999, 12, 28).atStartOfDay()
```
Comment (andrewwa-kainos) [inline src/test/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapperTest.java:113]: My usual points, that this test does not make it clear why these values should be expected without the reader having to leave the method. Do `all` patients get the same `NHS_NUMBER`?

Can we at least amend the `stubXXX()` methods to accept the values you'll be verifying later?

Similarly, it's not immediately obvious why `BIRTH_DATE_YYMMDD` and `CCYYMMDD` should map specifically to `1999`, `12`, `28`, but a little thought (without leaving the method!) lets us make a reasonable assumption quickly.

**Review by andrewwa-kainos — CHANGES_REQUESTED** [review]

**Review by saileshp-kainos — DISMISSED** [review]
LGTM

**Review by matthewdoreilly — DISMISSED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java (hunk 3 above)
Comment (n-r-b) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java]: Probably not 👍

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java (hunk 4 above)
Comment (n-r-b) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java]: Done

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java (hunk 5 above)
Comment (n-r-b) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java:75]: That could be possible! I'll re-write this slightly

File: src/test/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapperTest.java (hunk 6 above)
Comment (n-r-b) [reply src/test/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapperTest.java]: I've refactored the test cases, hopefully they make more sense. Let me know if you have any further suggestions

**Review by andrewwa-kainos — DISMISSED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java [hunk 7]
```diff
@@ -1,12 +1,83 @@
...
+import java.util.Optional;
 
 @Component
+@RequiredArgsConstructor
 public class PatientMapper {
+
+    protected static final String NHS_NUMBER_SYSTEM = "https://fhir.nhs.uk/Id/nhs-number";
+    private final UUIDGenerator uuidGenerator;
+
     public Patient mapToPatient(final Message message) {
```
Comment (saileshp-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PatientMapper.java:26]: I believe we're missing the address mapping `ADR++US` (`UnstructuredAddress.java`) in the `PatientMapper`

**Review by saileshp-kainos — CHANGES_REQUESTED** [review]

**Review by saileshp-kainos — APPROVED** [review]

**Review by andrewwa-kainos — APPROVED** [review]
LGTM

**Pull request merged**